interp: cannot handle fmt.Sprint()

Env:

```
$ tinygo version
tinygo version 0.8.0 linux/amd64
```

main.go:

```go
package main

import "fmt"

var Foo string

func main() {
	fmt.Println(Foo)
}

func init() {
	Foo = fmt.Sprint("foo")
}
```

Command:

```
tinygo build -o main.wasm -target wasm main.go
```

Result:

```
panic: todo: store

goroutine 1 [running]:
github.com/tinygo-org/tinygo/interp.(*Eval).hasLocalSideEffects(0xc0017ebce0, 0xc000e920a8, 0x7f183020e7a8, 0xc001098c00)
        /home/circleci/project/interp/scan.go:179 +0x1d2
github.com/tinygo-org/tinygo/interp.(*Eval).hasSideEffects(0xc0017ebce0, 0x7f183002b248, 0x7f183002b248)
        /home/circleci/project/interp/scan.go:106 +0x555
github.com/tinygo-org/tinygo/interp.(*Eval).hasSideEffects(0xc0017ebce0, 0x7f183002bfb8, 0x7f183002bfb8)
        /home/circleci/project/interp/scan.go:100 +0x4f2
github.com/tinygo-org/tinygo/interp.(*Eval).hasSideEffects(0xc0017ebce0, 0x7f1830032ef8, 0x7f1830032ef8)
        /home/circleci/project/interp/scan.go:100 +0x4f2
github.com/tinygo-org/tinygo/interp.(*Eval).hasSideEffects(0xc0017ebce0, 0x7f18300320e8, 0x7f18300320e8)
        /home/circleci/project/interp/scan.go:100 +0x4f2
github.com/tinygo-org/tinygo/interp.(*Eval).hasSideEffects(0xc0017ebce0, 0x7f1830032e08, 0x7f1830032e08)
        /home/circleci/project/interp/scan.go:100 +0x4f2
github.com/tinygo-org/tinygo/interp.(*Eval).hasSideEffects(0xc0017ebce0, 0x7f1830030868, 0x7f1830030868)
        /home/circleci/project/interp/scan.go:100 +0x4f2
github.com/tinygo-org/tinygo/interp.(*Eval).hasSideEffects(0xc0017ebce0, 0x7f183002f758, 0x7f183002f758)
        /home/circleci/project/interp/scan.go:100 +0x4f2
github.com/tinygo-org/tinygo/interp.(*Eval).hasSideEffects(0xc0017ebce0, 0x7f1830030688, 0x1)
        /home/circleci/project/interp/scan.go:100 +0x4f2
github.com/tinygo-org/tinygo/interp.(*frame).evalBasicBlock(0xc000e932b8, 0x7f1830f89790, 0x0, 0x0, 0x0, 0xc031326638, 0x10, 0x10, 0xc00239dd60, 0xa704c1, ...)
        /home/circleci/project/interp/frame.go:426 +0x4d5f
github.com/tinygo-org/tinygo/interp.(*Eval).function(0xc0017ebce0, 0x7f1830030598, 0xc000e933f8, 0x2, 0x2, 0xc0023a3e80, 0x7, 0x0, 0x0, 0x0, ...)
        /home/circleci/project/interp/interp.go:104 +0x208
github.com/tinygo-org/tinygo/interp.(*Eval).Function(...)
        /home/circleci/project/interp/interp.go:87
github.com/tinygo-org/tinygo/interp.Run(0x66817c0, 0x667fcc0, 0x0, 0x0, 0x0)
        /home/circleci/project/interp/interp.go:74 +0x681
main.Compile(0x7fffdb22ceec, 0x7, 0x7fffdb22ced5, 0x9, 0xc00011a4e0, 0xc000e93ed0, 0xc000e93d28, 0x0, 0x0)
        /home/circleci/project/main.go:148 +0xd02
main.Build(0x7fffdb22ceec, 0x7, 0x7fffdb22ced5, 0x9, 0x7fffdb22cee7, 0x4, 0xc00008bed0, 0xc0000ec310, 0xffffffffffffffff)
        /home/circleci/project/main.go:339 +0xe2
main.main()
        /home/circleci/project/main.go:702 +0xfd7
```

main2.go:

```go
package main

import "fmt"

var Foo = fmt.Sprint("foo")

func main() {
	fmt.Println(Foo)
}
```

Command:

```
tinygo build -o main2.wasm -target wasm main2.go
```

Result:

```
panic: interp: GEP on a constant

goroutine 1 [running]:
github.com/tinygo-org/tinygo/interp.(*LocalValue).GetElementPtr(0xc000dbb820, 0xc000dc0fe8, 0x2, 0x2, 0x353f000, 0xc000dbb810)
        /home/circleci/project/interp/values.go:107 +0x288
github.com/tinygo-org/tinygo/interp.(*frame).evalBasicBlock(0xc00262d2b8, 0x7f1cbcf89f00, 0x7f1cbcf89b80, 0x0, 0x0, 0x0, 0x0, 0xc0029d50e8, 0x1, 0x1, ...)
        /home/circleci/project/interp/frame.go:132 +0x1e86
github.com/tinygo-org/tinygo/interp.(*Eval).function(0xc002087890, 0x7f1cbc030598, 0xc00262d3f8, 0x2, 0x2, 0xc000dc0e50, 0x7, 0x0, 0x0, 0x0, ...)
        /home/circleci/project/interp/interp.go:104 +0x208
github.com/tinygo-org/tinygo/interp.(*Eval).Function(...)
        /home/circleci/project/interp/interp.go:87
github.com/tinygo-org/tinygo/interp.Run(0x4a28680, 0x4a26b80, 0x0, 0x0, 0x0)
        /home/circleci/project/interp/interp.go:74 +0x681
main.Compile(0x7ffe5b25ceec, 0x7, 0x7ffe5b25ced5, 0x9, 0xc0001224e0, 0xc00262ded0, 0xc00262dd28, 0x0, 0x0)
        /home/circleci/project/main.go:148 +0xd02
main.Build(0x7ffe5b25ceec, 0x7, 0x7ffe5b25ced5, 0x9, 0x7ffe5b25cee7, 0x4, 0xc000099ed0, 0xc0000f2290, 0xffffffffffffffff)
        /home/circleci/project/main.go:339 +0xe2
main.main()
        /home/circleci/project/main.go:702 +0xfd7
```

main3.go:

```go
package main

import "fmt"

var Foo string

func main() {
	Foo = fmt.Sprint("foo")
	fmt.Println(Foo)
}
```

Command:

```
tinygo build -o main3.wasm -target wasm main3.go
```

Result: OK.

--- a/interp/scan.go
+++ b/interp/scan.go
@@ -6,6 +6,21 @@ import (
 
 type sideEffectSeverity int
 
+func (severity sideEffectSeverity) String() string {
+	switch severity {
+	case sideEffectInProgress:
+		return "in progress"
+	case sideEffectNone:
+		return "none"
+	case sideEffectLimited:
+		return "limited"
+	case sideEffectAll:
+		return "all"
+	default:
+		return "unknown"
+	}
+}
+
 const (
 	sideEffectInProgress sideEffectSeverity = iota // computing side effects is in progress (for recursive functions)
 	sideEffectNone                                 // no side effects at all (pure)

--- a/interp/scan_test.go
+++ b/interp/scan_test.go
@@ -1,0 +1,87 @@
+package interp
+
+import (
+	"os"
+	"sort"
+	"testing"
+
+	"tinygo.org/x/go-llvm"
+)
+
+var scanTestTable = []struct {
+	name            string
+	severity        sideEffectSeverity
+	mentionsGlobals []string
+}{
+	{"returnsConst", sideEffectNone, nil},
+	{"returnsArg", sideEffectNone, nil},
+	{"externalCallOnly", sideEffectNone, nil},
+	{"externalCallAndReturn", sideEffectLimited, nil},
+	{"externalCallBranch", sideEffectLimited, nil},
+	{"readCleanGlobal", sideEffectNone, []string{"cleanGlobalInt"}},
+	{"readDirtyGlobal", sideEffectLimited, []string{"dirtyGlobalInt"}},
+	{"callFunctionPointer", sideEffectAll, []string{"functionPointer"}},
+}
+
+func TestScan(t *testing.T) {
+	t.Parallel()
+
+	// Read the input IR.
+	path := "testdata/scan.ll"
+	ctx := llvm.NewContext()
+	buf, err := llvm.NewMemoryBufferFromFile(path)
+	os.Stat(path) // make sure this file is tracked by `go test` caching
+	if err != nil {
+		t.Fatalf("could not read file %s: %v", path, err)
+	}
+	mod, err := ctx.ParseIR(buf)
+	if err != nil {
+		t.Fatalf("could not load module:\n%v", err)
+	}
+
+	// Check all to-be-tested functions.
+	for _, tc := range scanTestTable {
+		// Create an eval object, for testing.
+		e := &Eval{
+			Mod:          mod,
+			TargetData:   llvm.NewTargetData(mod.DataLayout()),
+			dirtyGlobals: map[llvm.Value]struct{}{},
+		}
+
+		// Mark some globals dirty, for testing.
+		e.markDirty(mod.NamedGlobal("dirtyGlobalInt"))
+
+		// Scan for side effects.
+		fn := mod.NamedFunction(tc.name)
+		if fn.IsNil() {
+			t.Errorf("scan test: could not find tested function %s in the IR", tc.name)
+			continue
+		}
+		result := e.hasSideEffects(fn)
+
+		// Check whether the result is what we expect.
+		if result.severity != tc.severity {
+			t.Errorf("scan test: function %s should have severity %s but it has %s", tc.name, tc.severity, result.severity)
+		}
+
+		// Check whether the mentioned globals match with what we'd expect.
+		mentionsGlobalNames := make([]string, 0, len(result.mentionsGlobals))
+		for global := range result.mentionsGlobals {
+			mentionsGlobalNames = append(mentionsGlobalNames, global.Name())
+		}
+		sort.Strings(mentionsGlobalNames)
+		globalsMismatch := false
+		if len(result.mentionsGlobals) != len(tc.mentionsGlobals) {
+			globalsMismatch = true
+		} else {
+			for i, globalName := range mentionsGlobalNames {
+				if tc.mentionsGlobals[i] != globalName {
+					globalsMismatch = true
+				}
+			}
+		}
+		if globalsMismatch {
+			t.Errorf("scan test: expected %s to mention globals %v, but it mentions globals %v", tc.name, tc.mentionsGlobals, mentionsGlobalNames)
+		}
+	}
+}

--- a/interp/testdata/scan.ll
+++ b/interp/testdata/scan.ll
@@ -1,0 +1,53 @@
+target datalayout = "e-m:e-i64:64-f80:128-n8:16:32:64-S128"
+target triple = "x86_64--linux"
+
+define i64 @returnsConst() {
+  ret i64 0
+}
+
+define i64 @returnsArg(i64 %arg) {
+  ret i64 %arg
+}
+
+declare i64 @externalCall()
+
+define i64 @externalCallOnly() {
+  %result = call i64 @externalCall()
+  ret i64 0
+}
+
+define i64 @externalCallAndReturn() {
+  %result = call i64 @externalCall()
+  ret i64 %result
+}
+
+define i64 @externalCallBranch() {
+  %result = call i64 @externalCall()
+  %zero = icmp eq i64 %result, 0
+  br i1 %zero, label %if.then, label %if.done
+
+if.then:
+  ret i64 2
+
+if.done:
+  ret i64 4
+}
+
+@cleanGlobalInt = global i64 5
+define i64 @readCleanGlobal() {
+  %global = load i64, i64* @cleanGlobalInt
+  ret i64 %global
+}
+
+@dirtyGlobalInt = global i64 5
+define i64 @readDirtyGlobal() {
+  %global = load i64, i64* @dirtyGlobalInt
+  ret i64 %global
+}
+
+@functionPointer = global i64()* null
+define i64 @callFunctionPointer() {
+  %fp = load i64()*, i64()** @functionPointer
+  %result = call i64 %fp()
+  ret i64 %result
+}

